File: FSharpTracer/FSharpTracer/Plane.fs
namespace FSharpTracer

type Plane = 
   { PlaneNormal : Vector
     Offset : float
     Surface : Surface }
   interface SceneObject with
      
      member p.Intersects ray = 
         let denom = p.PlaneNormal.Dot(ray.Direction)
         if denom > 0.0 then None
         else 
            Some({ Ray = ray
                   Distance = (p.PlaneNormal.Dot(ray.Position) + p.Offset) / -denom
                   Object = p })
      
      member p.Normal _ = p.PlaneNormal
      member p.GetSurface() = p.Surface

File: SwiftTracer/SwiftTracer/Plane.swift
import Foundation

class Plane : SceneObject {
    let planeNormal : Vector
    let offset : Double
    let surface : Surface
    init(planeNormal:Vector, offset:Double, surface:Surface) {
        self.planeNormal = planeNormal
        self.offset = offset
        self.surface = surface
    }
    
    func Intersects(ray: Ray) -> Optional<Intersection> {
        let denom = self.planeNormal.Dot(ray.direction)
        if denom > 0.0 {
            return .None
        } else {
            return .Some(Intersection(ray: ray, distance: (self.planeNormal.Dot(ray.position) + self.offset) / -denom, object: self))
        }
    }
    
    func Normal(_: Vector) -> Vector {
        return self.planeNormal
    }
    
    func GetSurface() -> Surface {
        return self.surface
    }
}